_source/_posts/2022-10-11-angular-dependency-injection.md

```diff
@@ -0,0 +1,332 @@
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class MyBreadComponent {
+  constructor(@Inject(BREAD_TOKEN) private bread: string) { }
+}
+```
+
+Now we can access the 🍞 emoji from within the component! This might not seem like a big deal, but having injection tokens as a means to represent values and interfaces as dependencies is enormous! And it sets us up to leverage the power of Angular's DI system.
+
+We can use injection tokens along with configuring providers within Angular's DI system for more power and fine-grained control.
+
+## Configuring providers in Angular's Dependency Injection system
+
+You can configure the `providers` array to add fine-grained control to your providers. When combined with injection tokens, we can unleash a lot of power. But first, it's essential to know when it makes sense to do so. Always prefer the most straightforward, default way of registering a dependency and then use fine-grained control as needed.
+
+To configure the `providers` array, you add an object containing the instructions like this:
+
+```ts
+@NgModule({
+  imports: // stuff here
+  declarations: // stuff here
+  providers: [
+    { provide: MyService, howToProvide: OtherDependency }
+  ]
+})
+export class AppModule {
+}
+```
+
+The "how to provide" gives Angular-specific instructions on this dependency configuration. Then you can provide the other new dependency. Angular supports the following options for "how to provide":
+
+1. `useClass` - Replace the current dependency with a new instance of something else
+2. `useExisting` - Replace the current dependency with an existing dependency
+3. `useValue` - Replace the current dependency with a new value
+4. `useFactory` - Use a factory method to determine which dependency to use based on a dynamic value
+
+Next, let's walk through examples of each configuration option to understand how to use them.
+
+### Configure providers with `useClass`
+
+The `useClass` option replaces the current dependency with a new instance of another class. This is a great option to use if you're refactoring code and want to substitute a different dependency in your application quickly. Let's say you have a language learning app and an Angular service that wraps the authentication calls you delegate to an auth library and an auth provider. We'll call this service `AuthService`, and to keep things straightforward, the code looks like this:
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  public login(): void { }
+  public logout(): void { }
+}
+```
+
+In a stroke of luck, a large tech company decides to buy your language learning app, but they require you to authenticate using their social login only. You can create a new authentication service; we'll call it `NewAuthService` (note, you should not name your services with these terrible names), and keep the same member names that wraps the calls to their auth provider. 
```

Comment:
```suggestion
In a stroke of luck, a large tech company decides to buy your language learning app, but they require you to authenticate using their social login only. You can create a new authentication service that wraps the calls to their auth provider and keeps the same member names; we'll call it `NewAuthService`. (Note, you should not name your services with these terrible generic names. Be a bit more descriptive. )
```
I tried to simplify the last long sentence because I found it confusing. Hopefully I didn't misunderstand it... n